// redux-form/es/actionTypes.js
export var ARRAY_INSERT = '@@redux-form/ARRAY_INSERT';
export var ARRAY_MOVE = '@@redux-form/ARRAY_MOVE';
export var ARRAY_POP = '@@redux-form/ARRAY_POP';
export var ARRAY_PUSH = '@@redux-form/ARRAY_PUSH';
export var ARRAY_REMOVE = '@@redux-form/ARRAY_REMOVE';
export var ARRAY_REMOVE_ALL = '@@redux-form/ARRAY_REMOVE_ALL';
export var ARRAY_SHIFT = '@@redux-form/ARRAY_SHIFT';
export var ARRAY_SPLICE = '@@redux-form/ARRAY_SPLICE';
export var ARRAY_UNSHIFT = '@@redux-form/ARRAY_UNSHIFT';
export var ARRAY_SWAP = '@@redux-form/ARRAY_SWAP';
export var AUTOFILL = '@@redux-form/AUTOFILL';
export var BLUR = '@@redux-form/BLUR';
export var CHANGE = '@@redux-form/CHANGE';
export var CLEAR_SUBMIT = '@@redux-form/CLEAR_SUBMIT';
export var DESTROY = '@@redux-form/DESTROY';
export var FOCUS = '@@redux-form/FOCUS';
export var INITIALIZE = '@@redux-form/INITIALIZE';
export var REGISTER_FIELD = '@@redux-form/REGISTER_FIELD';
export var RESET = '@@redux-form/RESET';
export var SET_SUBMIT_FAILED = '@@redux-form/SET_SUBMIT_FAILED';
export var SET_SUBMIT_SUCCEEDED = '@@redux-form/SET_SUBMIT_SUCCEEDED';
export var START_ASYNC_VALIDATION = '@@redux-form/START_ASYNC_VALIDATION';
export var START_SUBMIT = '@@redux-form/START_SUBMIT';
export var STOP_ASYNC_VALIDATION = '@@redux-form/STOP_ASYNC_VALIDATION';
export var STOP_SUBMIT = '@@redux-form/STOP_SUBMIT';
export var SUBMIT = '@@redux-form/SUBMIT';
export var TOUCH = '@@redux-form/TOUCH';
export var UNREGISTER_FIELD = '@@redux-form/UNREGISTER_FIELD';
export var UNTOUCH = '@@redux-form/UNTOUCH';
export var UPDATE_SYNC_ERRORS = '@@redux-form/UPDATE_SYNC_ERRORS';
export var UPDATE_SYNC_WARNINGS = '@@redux-form/UPDATE_SYNC_WARNINGS';
